Rework Pillar `top` matching and mine population

NicolasT opened the pull request.

NicolasT committed:
salt: rework Pillar `top.sls` matching

See: 7ca23367ae15c33e9f186c89f5a9615df541cba5
See: 6a23f45b0c3dde61283ab7ee66338de72f852a1e
See: c5b22bb688c54999f8be4907fc884e64ba17df46
See: #960
See: https://github.com/scality/metalk8s/issues/960

NicolasT committed:
salt: use `mine_functions` in Pillar

When using `mine.send` in states, the result is not necessarily
persistent (e.g. after `mine.flush`, a master cache loss,...). Instead
of running `mine.send` in various states only, we can define
`mine_functions` in targeted Pillars to update the mine when required,
or at regular intervals initiated by the master.

Fixes: #960
See: https://github.com/scality/metalk8s/issues/960

bert-e commented:
# Hello nicolast,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/960-mine-pillar` into
`development/2.0`
* [w/2.1/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/960-mine-pillar) will be merged into `development/2.1` 
* [w/2.2/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/960-mine-pillar) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

NicolasT committed:
salt: remove `online_mode`

This is an unsupported installation method, and completely untested, so
no use in keeping it around.

Fixes: #897
See: https://github.com/scality/metalk8s/issues/897

NicolasT committed:
salt: remove default for `registry_ip`

The default for `registry_ip`, `127.0.0.1`, is never actually 'valid' in
our deployments, hence it's not a reasonable fallback/default. Instead,
we should fail if no value is provided in the Pillar.

bert-e commented:
# History mismatch



Merge commit #756f42dca5e80f2c14fdd1f29ed7c9787dbd8138 on the integration branch
`w/2.1/improvement/960-mine-pillar` is merging a branch which is neither the current
branch `improvement/960-mine-pillar` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/960-mine-pillar` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/960-mine-pillar` into
`development/2.0`
* [w/2.1/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/960-mine-pillar) will be merged into `development/2.1` 
* [w/2.2/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/960-mine-pillar) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

bert-e commented:
# History mismatch



Merge commit #e58fe6c449f711e2bf723d720c57140319fb2db4 on the integration branch
`w/2.1/improvement/960-mine-pillar` is merging a branch which is neither the current
branch `improvement/960-mine-pillar` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/960-mine-pillar` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #77411834ad6400969d262567a47ed5c4e5d22cc0 on the integration branch
`w/2.1/improvement/960-mine-pillar` is merging a branch which is neither the current
branch `improvement/960-mine-pillar` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/960-mine-pillar` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/960-mine-pillar` into
`development/2.0`
* [w/2.1/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/960-mine-pillar) will be merged into `development/2.1` 
* [w/2.2/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/960-mine-pillar) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

TeddyAndrieux reviewed (commented):
Why did you not removed all the `mine.send` from the states ?

For me instead of `mine.send` we should only call `mine.update` it the state just after CAs certificates generations

NicolasT commented:
> Why did you not removed all the `mine.send` from the states ?
> 
> For me instead of `mine.send` we should only call `mine.update` it the state just after CAs certificates generations

I tried that at first. It doesn't work: the pillar matching relies on roles, and roles are only available once the API server is up, which is not the case when we need access to the certificates for the first time, when bootstrapping the bootstrap node. So, even when running `mine.update`, the `mine_functions` would not be properly set, so no values are pushed, and later `mine.get`s fail.

So, I kept the `mine.send`s for bootstrap sake only.

bert-e commented:
# History mismatch



Merge commit #c257b09e52269c8d297492489a2b58e76623cc19 on the integration branch
`w/2.1/improvement/960-mine-pillar` is merging a branch which is neither the current
branch `improvement/960-mine-pillar` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/960-mine-pillar` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/960-mine-pillar` into
`development/2.0`
* [w/2.1/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/960-mine-pillar) will be merged into `development/2.1` 
* [w/2.2/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/960-mine-pillar) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

bert-e commented:
# History mismatch



Merge commit #7a6c6590549c34d3d93f52f1f64618571781c9e9 on the integration branch
`w/2.1/improvement/960-mine-pillar` is merging a branch which is neither the current
branch `improvement/960-mine-pillar` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/960-mine-pillar` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT committed:
salt: silence `metalk8s_endpoints` `ext_pillar`

If the endpoints this `ext_pillar` looks up doesn't exist (yet), there's
no need to log an exception.

See: #984
See: https://github.com/scality/metalk8s/pull/984
See: #896
See: https://github.com/scality/metalk8s/issues/896

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/960-mine-pillar` into
`development/2.0`
* [w/2.1/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/960-mine-pillar) will be merged into `development/2.1` 
* [w/2.2/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/960-mine-pillar) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

NicolasT commented:
@bert-e approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf










*The following options are set:* **approve**

TeddyAndrieux commented on pillar/top.sls.in:
```diff
@@ -1,5 +1,33 @@
+  # the API server and associated `ext_pillar`s are in place).
+  # To work-around this, we assign the 'bootstrap' role Pillar data to all nodes
+  # for now, until either a better solution is found, or we get rid of
+  # `iso_root_path` and instead use what comes out of #924.
```
We can also in the bootstrap script create a pillar file only for the current node assigning this bootstrap roles.

Have something like that in the top.sls
```
metalk8s-{{ version }}:
  '*':
    - {{ grains.id }}
   [...]
```
And then in our bootstrap script we just do 
```echo "{'metalk8s': {'nodes': {'$bootstrap_id': {'roles': ['ca', 'bootstrap', 'master', 'etcd']}}}"```
But then we have an issue with `ext_pillar_first` as this information in the static file will be take over the one in the apiserver...

So I don't really know.
Not really like the `ext_pillar_first` as the information from the apiserver should be taken over the one in the pillar file.

We should maybe get rid of "classic" pillar and only use ext_pillar to set all the information including the mine_functions... and so on.
And then in the master config `ext_pillar` we set the list in the right order we want them to get merged and then with "pillar" args from `def ext_pillar(minion_id, pillar)` we can get the information like roles and also the one from the command line.

Just tested `pillar` arg from ext_pillar contain all the merged informations of the previous ext_pillar + static pillar + **cmd line pillar** so for me it's something we can do

NicolasT replied:
> We can also in the bootstrap script create a pillar file only for the current node assigning this bootstrap roles.

Could even be simpler: pass it as a pillar value on the `salt-call` invocation in the script, like we do for several other pillar values.

NicolasT replied:
>  Not really like the ext_pillar_first as the information from the apiserver should be taken over the one in the pillar file.

Yes. And no. Not using `ext_pillar_first` simply doesn't work: we require the `ext_pillar` values to be able to dispatch on the roles (which are in the pillar...). Basically, when you don't use `ext_pillar_first`), you get crashes on the master side when it tries to compile the pillar top for a minion if said pillar top SLS contains matches on pillar values. So, no way around.

NicolasT replied:
> We should maybe get rid of "classic" pillar and only use ext_pillar to set all the information including the mine_functions... and so on.

Nah. If we can have static data, let's use static data.

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/960-mine-pillar` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`improvement/960-mine-pillar`).

```bash
 $ git fetch
 $ git checkout origin/improvement/960-mine-pillar
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/960-mine-pillar
```







*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #3db695c358534960b68f59387783c58da2ed3030 on the integration branch
`w/2.1/improvement/960-mine-pillar` is merging a branch which is neither the current
branch `improvement/960-mine-pillar` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/960-mine-pillar` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/960-mine-pillar` into
`development/2.0`
* [w/2.1/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/960-mine-pillar) will be merged into `development/2.1` 
* [w/2.2/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/960-mine-pillar) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf










*The following options are set:* **approve**

TeddyAndrieux reviewed (dismissed):
LGTM

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/6e6083eda86004d68c5cfd296b19194083157fdb) did not succeed in branch improvement/960-mine-pillar.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #50e7ef0ea6208da6734371f27b9fb0ffca6faaac on the integration branch
`w/2.1/improvement/960-mine-pillar` is merging a branch which is neither the current
branch `improvement/960-mine-pillar` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/960-mine-pillar` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/960-mine-pillar` into
`development/2.0`
* [w/2.1/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/960-mine-pillar) will be merged into `development/2.1` 
* [w/2.2/improvement/960-mine-pillar](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/960-mine-pillar) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/91bf41c59e5754b9dbfb274f895321614109dbac) did not succeed in branch w/2.1/improvement/960-mine-pillar.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye nicolast.
